File: back/runtime/doltgresspike/doltgresspike.go
// Package doltgresspike is the S88 gating spike (EPIC 9, ADR 0006 addendum).
//
// It answers ONE governance question deterministically: does the emitted app's
// data-versioning target default to Doltgres, or to plain-Postgres?
//
// Per ADR 0006 the open question was "is Doltgres the default emitted-app
// datastore, or opt-in alongside plain-Postgres?". This spike resolves it with a
// MEASURE, never an opinion (§8 "the judge is deterministic", determinism-first
// mandate). A Testcontainers probe (doltgresspike_concurrency_test.go) drives N
// concurrent connections at the emitted app's data driver against a Doltgres
// container and records a Measurement; this package turns that Measurement into a
// content-addressed Decision via a PURE function — same Measurement + same
// declared Thresholds → byte-identical Decision (reproducibility mirror).
//
// The DEFAULT is plain-Postgres and Doltgres is OPT-IN per app. The spike can
// only CONFIRM that default (go: Doltgres is also offerable as opt-in) or HARDEN
// it (no-go: Doltgres is withdrawn even as opt-in until the beta caveats clear).
// Either way the escape hatch (plain-Postgres) is already the default, so EPIC 10
// (S91→S100) stays viable without reordering — exactly as the roadmap requires.
//
// EPIC 9 preamble note (roadmap): the emitted app's data driver is a TypeScript
// Postgres client (postgres.js / node-postgres / Drizzle), NOT pgx. The pgx-only
// failure mode #2581 (thread-unsafe parser under concurrent pgx) may not
// reproduce identically against a TS driver, so the go/no-go criteria are
// re-formulated per the TS driver here: the probe measures connection STABILITY
// (no panic / no dropped connection / no corrupted result) and the perf RATIO
// against plain-Postgres (the ~5.2× and missing lock primitives #2600 are
// dialect/engine-level and survive the driver pivot).
package doltgresspike

import (
	"encoding/json"
	"fmt"
	"sort"

	"github.com/steph-frtech/aidos/back/kernel/records"
)

// Target is an emitted-app datastore target (ADR 0006 frozen-stack slot,
// replaceable). PlainPostgres is the default; Doltgres is opt-in.
type Target string

const (
	// PlainPostgres is the DEFAULT emitted-app datastore target. It is always
	// available regardless of the spike verdict (the escape hatch).
	PlainPostgres Target = "plain-postgres"
	// Doltgres is the opt-in, git-for-data emitted-app datastore target. It is
	// only OFFERED (as opt-in) when the spike verdict is Go.
	Doltgres Target = "doltgres"
)

// Verdict is the spike's binary gate. It is COMPUTED from a Measurement against
// declared Thresholds, never declared by an agent.
type Verdict string

const (
	// Go means Doltgres survived the concurrency + load probe within the declared
	// thresholds; it MAY be offered as an opt-in target. The default stays
	// plain-Postgres regardless.
	Go Verdict = "go"
	// NoGo means the probe found a reproducible instability or the perf ratio blew
	// past the declared ceiling; Doltgres is withdrawn even as opt-in until the
	// beta caveats clear. The default (already plain-Postgres) is unchanged.
	NoGo Verdict = "no-go"
)

// Driver names the data driver the probe exercised. Per the EPIC 9 preamble the
// emitted-app primary is a TS driver; pgx is retained only as the AIDOS-side
// equivalent stand-in used by the Testcontainers probe when no TS runner is
// available (the failure mode is re-formulated, not the engine under test).
type Driver string

const (
	// DriverTSPostgres is the emitted-app primary driver (postgres.js /
	// node-postgres / Drizzle adapter) — the one S88 must judge against.
	DriverTSPostgres Driver = "ts-postgres"
	// DriverPgx is the Go-side stand-in the in-repo Testcontainers probe uses; its
	// #2581 failure mode is pgx-specific and re-formulated to driver-neutral
	// STABILITY for the verdict.
	DriverPgx Driver = "pgx"
)

// Thresholds are the DECLARED go/no-go criteria (above the line, §8 "weights
// declared, never learned"). They are part of the verdict's content-address, so a
// changed threshold yields a different Decision hash — the verdict can never be
// silently re-graded.
type Thresholds struct {
	// MaxPerfRatio is the largest tolerated (Doltgres latency / plain-Postgres
	// latency) under the concurrent load. ADR 0006 cites ~5.2× as the known
	// dialect-level slowdown; the spike fails (no-go) only above the declared
	// ceiling, so a result merely confirming ~5.2× is still a Go.
	MaxPerfRatio float64 `json:"maxPerfRatio"`
	// MaxFailedConns is the largest tolerated number of connections that ended in
	// a panic / dropped connection / corrupted result across the probe. Default 0:
	// any reproducible instability is a no-go.
	MaxFailedConns int `json:"maxFailedConns"`
}

// DefaultThresholds are the declared S88 criteria. MaxPerfRatio 6.0 leaves head-
// room above ADR 0006's cited ~5.2× (a result near the documented slowdown is a
// Go, a runaway is a no-go); MaxFailedConns 0 means a single reproducible
// instability flips the default. These are DECLARED, not tuned to a result.
var DefaultThresholds = Thresholds{MaxPerfRatio: 6.0, MaxFailedConns: 0}

// Measurement is the raw, deterministic OUTPUT of the Testcontainers probe — a
// pure observation, no judgment. The probe fills it; this package judges it.
type Measurement struct {
	// Driver is the data driver exercised.
	Driver Driver `json:"driver"`
	// Conns is N, the number of concurrent connections submitted.
	Conns int `json:"conns"`
	// FailedConns is how many ended in panic / dropped connection / corrupted
	// result (the driver-neutral re-formulation of #2581 / #2600).
	FailedConns int `json:"failedConns"`
	// PerfRatio is the measured (Doltgres / plain-Postgres) latency ratio under
	// the concurrent load. 0 means not measured (treated as worst case below).
	PerfRatio float64 `json:"perfRatio"`
	// Reproducible reports whether the same probe re-run produced the same
	// failure/stability outcome. A non-reproducible blip is NOT a no-go (the
	// done-criteria require a REPRODUCIBLE failure to flip the default).
	Reproducible bool `json:"reproducible"`
}

// Decision is the content-addressed record the spike emits. It is a RECORD, never
// a declaration: ID = Hash(canonical-json of the inputs+verdict), so the same
// measurement always yields the same Decision and the verdict is auditable.
type Decision struct {
	// ID is the content address (records.Hash of the canonical payload).
	ID string `json:"id"`
	// Verdict is the computed gate.
	Verdict Verdict `json:"verdict"`
	// DefaultTarget is the emitted-app datastore default after this decision —
	// ALWAYS plain-postgres (the escape hatch is the default by construction).
	DefaultTarget Target `json:"defaultTarget"`
	// OptInTargets are the targets an app may opt into. plain-postgres is always
	// present; doltgres is present iff the verdict is Go.
	OptInTargets []Target `json:"optInTargets"`
	// Measurement is the raw observation that produced the verdict.
	Measurement Measurement `json:"measurement"`
	// Thresholds are the declared criteria the measurement was judged against.
	Thresholds Thresholds `json:"thresholds"`
	// Reasons explains the verdict in actionable, deterministic terms.
	Reasons []string `json:"reasons"`
}

// Evaluate is the AUTHORITATIVE pure function: Measurement × Thresholds → Verdict.
// It is total, deterministic and side-effect-free (the reproducibility mirror
// pins same-input→same-output). No LLM may produce this verdict — it is a count
// and two comparisons.
//
// A measurement is a no-go iff a REPRODUCIBLE instability is found beyond the
// declared connection ceiling, OR the perf ratio blows past the declared ceiling.
// A non-reproducible blip never flips the default (the done-criteria demand a
// reproducible failure). An unmeasured perf ratio (0) is treated as worst case.
func Evaluate(m Measurement, th Thresholds) (Verdict, []string) {
	reasons := make([]string, 0, 3)

	stableEnough := m.FailedConns <= th.MaxFailedConns
	if !stableEnough && m.Reproducible {
		reasons = append(reasons, fmt.Sprintf(
			"reproducible instability: %d/%d connections failed (ceiling %d) under driver %q",
			m.FailedConns, m.Conns, th.MaxFailedConns, m.Driver))
	}
	if !stableEnough && !m.Reproducible {
		// Instability seen but not reproducible: per done-criteria this does NOT
		// flip the default; it is recorded as a caveat, not a no-go.
		reasons = append(reasons, fmt.Sprintf(
			"non-reproducible instability (%d/%d) — NOT a no-go per done-criteria; recorded as caveat",
			m.FailedConns, m.Conns))
		stableEnough = true
	}

	ratio := m.PerfRatio
	perfOK := ratio > 0 && ratio <= th.MaxPerfRatio
	if ratio <= 0 {
		reasons = append(reasons, "perf ratio not measured — treated as worst case (no-go)")
	} else if !perfOK {
		reasons = append(reasons, fmt.Sprintf(
			"perf ratio %.2f× exceeds declared ceiling %.2f×", ratio, th.MaxPerfRatio))
	} else {
		reasons = append(reasons, fmt.Sprintf(
			"perf ratio %.2f× within declared ceiling %.2f×", ratio, th.MaxPerfRatio))
	}

	if stableEnough && perfOK {
		reasons = append(reasons, fmt.Sprintf(
			"stability OK: %d/%d connections survived under driver %q → Doltgres offerable as opt-in",
			m.Conns-m.FailedConns, m.Conns, m.Driver))
		return Go, reasons
	}
	return NoGo, reasons
}

// Decide turns a Measurement into a content-addressed Decision. The default
// target is ALWAYS plain-postgres (the escape hatch by construction); Doltgres
// joins the opt-in set only on a Go verdict. The ID is the records.Hash of the
// canonical (verdict-independent) payload so the record is reproducible and
// tamper-evident.
func Decide(m Measurement, th Thresholds) Decision {
	verdict, reasons := Evaluate(m, th)

	optIn := []Target{PlainPostgres}
	if verdict == Go {
		optIn = append(optIn, Doltgres)
	}
	sortTargets(optIn)

	d := Decision{
		Verdict:       verdict,
		DefaultTarget: PlainPostgres, // escape hatch is the default, always.
		OptInTargets:  optIn,
		Measurement:   m,
		Thresholds:    th,
		Reasons:       reasons,
	}
	d.ID = records.Hash(d.canonicalPayload())
	return d
}

// canonicalPayload is the deterministic byte representation hashed into the ID.
// It excludes the ID itself and the human-readable Reasons (which are derived,
// not authoritative), so the content-address keys ONLY on the load-bearing
// inputs+verdict — two runs that agree on those agree on the ID.
func (d Decision) canonicalPayload() []byte {
	type payload struct {
		Verdict       Verdict     `json:"verdict"`
		DefaultTarget Target      `json:"defaultTarget"`
		OptInTargets  []Target    `json:"optInTargets"`
		Measurement   Measurement `json:"measurement"`
		Thresholds    Thresholds  `json:"thresholds"`
	}
	b, _ := json.Marshal(payload{
		Verdict:       d.Verdict,
		DefaultTarget: d.DefaultTarget,
		OptInTargets:  d.OptInTargets,
		Measurement:   d.Measurement,
		Thresholds:    d.Thresholds,
	})
	return b
}

// OptInAllowed reports whether an app may opt into the given target after this
// decision. plain-postgres is always allowed; doltgres only on a Go verdict.
func (d Decision) OptInAllowed(t Target) bool {
	for _, ot := range d.OptInTargets {
		if ot == t {
			return true
		}
	}
	return false
}

func sortTargets(ts []Target) {
	sort.Slice(ts, func(i, j int) bool { return ts[i] < ts[j] })
}

File: front/web/lib/doltgres-spike.ts
/**
 * doltgres-spike — the TS twin of back/runtime/doltgresspike (S88, ADR 0006
 * addendum 0047). DETERMINISM-FIRST (CLAUDE.md §6/§8): the spike verdict is a
 * PURE function — a count + two comparisons, NEVER an LLM. Same Measurement +
 * Thresholds → byte-identical Decision (same content-addressed id). The Vitest
 * mirror (doltgres-spike.test.ts) pins same-input→same-output and the gating
 * invariants; the authoritative engine is the Go package (this twin lets the
 * Workbench run the verdict from the screen without a backend round-trip).
 *
 * THE WALL (CLAUDE.md §2): pure judgment over a supplied measurement — writes
 * NOTHING. The emitted-app DEFAULT target is ALWAYS plain-postgres (the escape
 * hatch by construction); Doltgres is opt-in iff the verdict is Go.
 */

export type Target = "plain-postgres" | "doltgres";
export type Verdict = "go" | "no-go";
export type Driver = "ts-postgres" | "pgx";

export interface Thresholds {
	/** largest tolerated (Doltgres / plain-Postgres) latency ratio under load. */
	maxPerfRatio: number;
	/** largest tolerated number of failed connections (panic/drop/corruption). */
	maxFailedConns: number;
}

/** the DECLARED S88 criteria (above the line, never learned). */
export const DEFAULT_THRESHOLDS: Thresholds = {
	maxPerfRatio: 6.0,
	maxFailedConns: 0,
};

export interface Measurement {
	driver: Driver;
	/** N — concurrent connections submitted. */
	conns: number;
	/** how many ended in panic / dropped connection / corrupted result. */
	failedConns: number;
	/** measured (Doltgres / plain-Postgres) latency ratio; 0 = not measured. */
	perfRatio: number;
	/** whether a re-run produced the same outcome (a blip never flips the default). */
	reproducible: boolean;
}

export interface Decision {
	id: string;
	verdict: Verdict;
	/** ALWAYS plain-postgres — the escape hatch by construction. */
	defaultTarget: Target;
	/** plain-postgres always; doltgres iff verdict === "go". */
	optInTargets: Target[];
	measurement: Measurement;
	thresholds: Thresholds;
	reasons: string[];
}

/**
 * evaluate is the authoritative pure judgment: Measurement × Thresholds →
 * (verdict, reasons). Total, deterministic, side-effect-free. A no-go requires a
 * REPRODUCIBLE instability past the connection ceiling, OR a perf ratio over the
 * declared ceiling; a non-reproducible blip never flips the default (the
 * done-criteria require a reproducible failure). An unmeasured ratio (0) is the
 * worst case.
 */
export function evaluate(
	m: Measurement,
	th: Thresholds,
): { verdict: Verdict; reasons: string[] } {
	const reasons: string[] = [];

	let stableEnough = m.failedConns <= th.maxFailedConns;
	if (!stableEnough && m.reproducible) {
		reasons.push(
			`reproducible instability: ${m.failedConns}/${m.conns} connections failed (ceiling ${th.maxFailedConns}) under driver "${m.driver}"`,
		);
	}
	if (!stableEnough && !m.reproducible) {
		reasons.push(
			`non-reproducible instability (${m.failedConns}/${m.conns}) — NOT a no-go per done-criteria; recorded as caveat`,
		);
		stableEnough = true;
	}

	const ratio = m.perfRatio;
	const perfOK = ratio > 0 && ratio <= th.maxPerfRatio;
	if (ratio <= 0) {
		reasons.push("perf ratio not measured — treated as worst case (no-go)");
	} else if (!perfOK) {
		reasons.push(
			`perf ratio ${ratio.toFixed(2)}× exceeds declared ceiling ${th.maxPerfRatio.toFixed(2)}×`,
		);
	} else {
		reasons.push(
			`perf ratio ${ratio.toFixed(2)}× within declared ceiling ${th.maxPerfRatio.toFixed(2)}×`,
		);
	}

	if (stableEnough && perfOK) {
		reasons.push(
			`stability OK: ${m.conns - m.failedConns}/${m.conns} connections survived under driver "${m.driver}" → Doltgres offerable as opt-in`,
		);
		return { verdict: "go", reasons };
	}
	return { verdict: "no-go", reasons };
}

/** the canonical bytes hashed into the decision id (twin of the Go canonicalPayload). */
export function canonicalPayload(
	verdict: Verdict,
	defaultTarget: Target,
	optInTargets: Target[],
	m: Measurement,
	th: Thresholds,
): string {
	// Matches Go's json.Marshal field order for byte-identical hashing.
	return JSON.stringify({
		verdict,
		defaultTarget,
		optInTargets,
		measurement: {
			driver: m.driver,
			conns: m.conns,
			failedConns: m.failedConns,
			perfRatio: m.perfRatio,
			reproducible: m.reproducible,
		},
		thresholds: {
			maxPerfRatio: th.maxPerfRatio,
			maxFailedConns: th.maxFailedConns,
		},
	});
}

/**
 * decide turns a Measurement into a content-addressed Decision. The default
 * target is ALWAYS plain-postgres; Doltgres joins the opt-in set only on a Go.
 *
 * The sha256 is imported LAZILY (dynamic `node:crypto`) so this module stays
 * client-bundle-safe — a client component may import the types + DEFAULT_THRESHOLDS
 * + MEASURED_SPIKE without dragging Node crypto into the browser bundle (the same
 * pattern as lib/relation-emitter's actions). `decide` therefore runs server-side
 * (Server Action) or in Node (Vitest), never in the browser.
 */
export async function decide(
	m: Measurement,
	th: Thresholds = DEFAULT_THRESHOLDS,
): Promise<Decision> {
	const { verdict, reasons } = evaluate(m, th);
	const optInTargets: Target[] =
		verdict === "go" ? ["doltgres", "plain-postgres"] : ["plain-postgres"];
	optInTargets.sort();
	const { createHash } = await import("node:crypto");
	const id = createHash("sha256")
		.update(canonicalPayload(verdict, "plain-postgres", optInTargets, m, th))
		.digest("hex");
	return {
		id,
		verdict,
		defaultTarget: "plain-postgres",
		optInTargets,
		measurement: m,
		thresholds: th,
		reasons,
	};
}

export function optInAllowed(d: Decision, t: Target): boolean {
	return d.optInTargets.includes(t);
}

/**
 * The measured S88 run (2026-06-08, decision 7395fa69c2d7): N=64 concurrent
 * connections against a real Doltgres container, 0 failures, reproducible, perf
 * within the declared ceiling → Go. Plain-postgres stays the default. Used to
 * pre-fill the screen so the control is reachable AND executable.
 */
export const MEASURED_SPIKE: Measurement = {
	driver: "pgx",
	conns: 64,
	failedConns: 0,
	perfRatio: 0.3,
	reproducible: true,
};
